fvm/transactionStorageLimiter.go

```diff
@@ -33,35 +44,70 @@ func (d TransactionStorageLimiter) CheckStorageLimits(
 
 	defer env.StartChildSpan(trace.FVMTransactionStorageUsedCheck).End()
 
-	err := d.checkStorageLimits(env, addresses, payer, maxTxFees)
+	err := limiter.checkStorageLimits(env, txnState, payer, maxTxFees)
 	if err != nil {
 		return fmt.Errorf("storage limit check failed: %w", err)
 	}
 	return nil
 }
 
-func (d TransactionStorageLimiter) checkStorageLimits(
-	env environment.Environment,
-	addresses []flow.Address,
+// getStorageCheckAddresses returns a list of addresses to be checked whether
+// storage limit is exceeded.  The returned list include addresses of updated
+// registers (and the payer's address).
+func (limiter TransactionStorageLimiter) getStorageCheckAddresses(
+	txnState *state.TransactionState,
 	payer flow.Address,
 	maxTxFees uint64,
-) error {
-	// in case the payer is not already part of the check, include it here.
-	// If the maxTxFees is zero, it doesn't matter if the payer is included or not.
+) []flow.Address {
+	updatedIds := txnState.UpdatedRegisterIDs()
+
+	// Multiple updated registers might be from the same address.  We want to
+	// duplicated the addresses to reduce check overhead.
+	dedup := make(map[flow.Address]struct{}, len(updatedIds)+1)
```

Comment:
The value of `len(updatedIds)` is always larger or equal to the accounts updated, but might be a lot larger, because you generally touch more registers per address.

In this instance it might be better to just create the map (and array) without a capacity. Not sure though...

Reply:
it's ok to over allocate since both the map and slice have short life span, and the memory will be reclaimed almost immediately.  map resizing would cause more gc overhead